packages/safe-ds-lang/src/language/generation/safe-ds-python-generator.ts

```diff
@@ -381,20 +394,180 @@ export class SafeDsPythonGenerator {
         )!;
     }
 
+    private getStatementsNeededForPartialExecution(
+        targetPlaceholder: SdsPlaceholder,
+        statementsWithEffect: SdsStatement[],
+    ): SdsStatement[] {
+        const neededPlaceholders = new Set<string>();
+        const neededStatements = new Set<SdsStatement>();
+        this.getAllPlaceholdersAndStatementsRelevantForPlaceholder(
+            targetPlaceholder,
+            statementsWithEffect,
+            neededPlaceholders,
+            neededStatements,
+        );
+        // Get all statements in sorted order
+        return Array.from(neededStatements.values()).sort((a, b) => a.$containerIndex! - b.$containerIndex!);
+    }
+
+    private getAllPlaceholdersAndStatementsRelevantForPlaceholder(
+        targetPlaceholder: SdsPlaceholder,
+        allStatements: SdsStatement[],
+        neededPlaceholders: Set<string>,
+        neededStatements: Set<SdsStatement>,
+    ) {
+        // If placeholder is inside a lambda, do not continue; Only placeholders inside the pipeline are collected
+        if (this.isNodeInsideBlockLambda(targetPlaceholder)) {
+            return;
+        }
+        // Find assignment of placeholder, to search used placeholders and impure dependencies
+
+        let assignment = getContainerOfType(targetPlaceholder, isSdsAssignment);
+        if (!assignment || !assignment.expression) {
+            /* c8 ignore next 2 */
+            return;
+        }
+        // If placeholder is already found, do not continue; It has been handled before
+        if (neededPlaceholders.has(targetPlaceholder.name) && neededStatements.has(assignment)) {
```

Comment:
This seems to be the only time `neededPlaceholders` is ever read and for this check the second condition `neededStatements.has(assignment)` seems to be sufficient.

Unless, there's another reason to keep `neededPlaceholders` around, I'd suggest to
1. Remove the parameter `neededPlaceholders`.
2. Make the `neededStatements` parameter the return value.
3. Rename the method `getStatementsRelevantFor...` (or `keepStatementsRelevantFor`).

This applies to `getAllPlaceholdersAndStatementsRelevantForPlaceholder` and `getAllPlaceholdersAndStatementsRelevantForStatement`.

-----

Generally, the logic in and interaction between
* `getAllPlaceholdersAndStatementsRelevantForPlaceholder`,
* `getAllPlaceholdersAndStatementsRelevantForStatement`,
* `getImpureDependenciesForStatement`

seems somewhat convoluted.

Isn't it possible to
1. Get the assignment containing the target placeholder (we always must evaluate this one).
2. Collect all placeholder references & impurity reasons in its expression.
3. Iteratively traverse the previous statements in reverse order and for each
    a. Check whether it defines a needed placeholder or interact with some known impurity reason. 
    b. If neither applies, we can omit it and continue with the previous one.
    c. Otherwise, we must generate code for it and collect all placeholder references & impurity reasons in its expression.